frame/balances/src/lib.rs

```diff
@@ -608,1602 +466,611 @@ pub mod pallet {
-	}
-
-	fn reactivate(amount: Self::Balance) {
-		<Self as fungible::Transfer<T::AccountId>>::reactivate(amount);
-	}
-
-	fn minimum_balance() -> Self::Balance {
-		T::ExistentialDeposit::get()
-	}
+		/// Mutate an account to some new value, or delete it entirely with `None`. Will enforce
+		/// `ExistentialDeposit` law, annulling the account as needed.
+		///
+		/// It returns both the result from the closure, and an optional amount of dust
+		/// which should be handled once it is known that all nested mutates that could affect
+		/// storage items what the dust handler touches have completed.
+		///
+		/// NOTE: Doesn't do any preparatory work for creating a new account, so should only be used
+		/// when it is known that the account already exists.
+		///
+		/// NOTE: LOW-LEVEL: This will not attempt to maintain total issuance. It is expected that
+		/// the caller will do this.
+		pub fn mutate_account<R>(
+			who: &T::AccountId,
+			f: impl FnOnce(&mut AccountData<T::Balance>) -> R,
+		) -> Result<(R, Option<T::Balance>), DispatchError> {
+			Self::try_mutate_account(who, |a, _| -> Result<R, DispatchError> { Ok(f(a)) })
+		}
+
+		/// Mutate an account to some new value, or delete it entirely with `None`. Will enforce
+		/// `ExistentialDeposit` law, annulling the account as needed. This will do nothing if the
+		/// result of `f` is an `Err`.
+		///
+		/// It returns both the result from the closure, and an optional amount of dust
+		/// which should be handled once it is known that all nested mutates that could affect
+		/// storage items what the dust handler touches have completed.
+		///
+		/// NOTE: Doesn't do any preparatory work for creating a new account, so should only be used
+		/// when it is known that the account already exists.
+		///
+		/// NOTE: LOW-LEVEL: This will not attempt to maintain total issuance. It is expected that
+		/// the caller will do this.
+		pub(crate) fn try_mutate_account<R, E: From<DispatchError>>(
+			who: &T::AccountId,
+			f: impl FnOnce(&mut AccountData<T::Balance>, bool) -> Result<R, E>,
+		) -> Result<(R, Option<T::Balance>), E> {
+			Self::ensure_upgraded(who);
+			let result = T::AccountStore::try_mutate_exists(who, |maybe_account| {
+				let is_new = maybe_account.is_none();
+				let mut account = maybe_account.take().unwrap_or_default();
+				let did_provide = account.free >= T::ExistentialDeposit::get();
+				let did_consume =
+					!is_new && (!account.reserved.is_zero() || !account.frozen.is_zero());
+
+				let result = f(&mut account, is_new)?;
+
+				let does_provide = account.free >= T::ExistentialDeposit::get();
+				let does_consume = !account.reserved.is_zero() || !account.frozen.is_zero();
+
+				if !did_provide && does_provide {
+					frame_system::Pallet::<T>::inc_providers(who);
+				}
+				if did_consume && !does_consume {
+					frame_system::Pallet::<T>::dec_consumers(who);
+				}
+				if !did_consume && does_consume {
+					frame_system::Pallet::<T>::inc_consumers(who)?;
+				}
```

Comment:
Having funds on hold/reserved or having freezes/locks active requires a consumer reference since it uses additional storage in the Balances pallet.

Staking also uses up storage so it requires a consumer reference. As it happens, staking also uses locks (soon to be updated to a hold), so it ends up using two references. But the reason for those two references are both valid and independent.

Reply:
Understood!